README.md
[![License: MIT](https://img.shields.io/badge/license-MIT-blue.svg?style=flat)](https://github.com/gunnarhafdal/GHNavigationDropdownMenu/blob/master/LICENSE)
![Language](https://img.shields.io/badge/language-Swift-brightgreen.svg?style=flat)

## Introduction:
The elegant dropdown menu, written in **Swift**, appears underneath navigation bar to display a list of related items when a user click on the navigation title. Forked from [PhamBaTho/BTNavigationDropdownMenu](https://github.com/PhamBaTho/BTNavigationDropdownMenu) because I needed Swift3 support and support for a subtitle in the menu items.

## Usage:
### Instantiating
Start by creating an Array that contains strings as **elements of dropdown list**:
```swift
let items = [
            GHDropdownItem(title: "Most Popular", subtitle: "from our users"),
            GHDropdownItem(title: "Latest", subtitle: ""),
            GHDropdownItem(title: "Trending", subtitle: "See what's happening"),
            GHDropdownItem(title: "Nearest", subtitle: "to you"),
            GHDropdownItem(title: "Top Picks", subtitle: "by our editors"),
        ]
```
Create a **new instance** of GHNavigationDropdownMenu:
```swift
let menuView = GHNavigationDropdownMenu(navigationController: self.navigationController, containerView: self.navigationController!.view, title: "Dropdown Menu", items: items)
```
or just simple like this:
```swift
let menuView = GHNavigationDropdownMenu(title: items[0], items: items)
```
By default, `navigationController` is the top most navigation controller and `containerView` is keyWindow. 

(`keyWindow` is recommended for `containerView` because in this way, the black overlay can cover the whole screen. But in some cases, `keyWindow` doesn't work properly, like using with side menu, (e.g. SWRevealViewController), the dropdown menu didn't move along with their parent view controller or navigation controller. To resolve this issue, you can use `self.navigationController!.view` instead.)

Set **title of navigation bar** as menuView:
```swift
self.navigationItem.titleView = menuView
```
Call GHNavigationDropdownMenu closure to get **the index of selected cell**:
```swift
menuView.didSelectItemAtIndexHandler = {[weak self] (indexPath: Int) -> () in
            print("Did select item at index: \(indexPath)")
            self.selectedCellLabel.text = items[indexPath]
}
```
Use `menuView.show()` or `menuView.hide()` if you want to show or hide dropdown menu manually.

Use `menuView.toggle()` to toogle dropdown menu shown/hide.

Use `menuView.isShown` (Boolean type) property to check showing state of dropdown menu.

Use `menuView.updateItems(items: [AnyObject])` to update items in dropdown menu if needed.

### Customization
Once you have assigned the items and frame for dropdown menu, you can custom the look and the feel of menu. You can override these properties for your favor:

`cellHeight` **- The height of the cell.** *Default is 50*

`cellBackgroundColor` **- The color of the cell background.** *Default is whiteColor()*

`cellSeparatorColor` **- The color of the cell separator.** *Default is darkGrayColor()*

`cellTextLabelColor` **- The color of the text inside cell.** *Default is darkGrayColor()*

`cellTextLabelFont` **- The font of the text inside cell.** *Default is HelveticaNeue-Bold, size 17*

`navigationBarTitleFont` **- The font of the navigation bar title.** *Default is HelveticaNeue-Bold, size 17*

`cellTextLabelAlignment` **- The alignment of the text inside cell.** *Default is .Left*

`cellSelectionColor`  **- The color of the cell when the cell is selected.** *Default is lightGrayColor()*

`checkMarkImage`  **- The checkmark icon of the cell.**

`animationDuration`  **- The animation duration of showing/hiding menu.** *Default is 0.5s*

`arrowImage`  **- The arrow next to navigation title**

`arrowPadding`  **- The padding between navigation title and arrow.** *Default is 15*

`maskBackgroundColor`  **- The color of the mask layer.** *Default is blackColor()*

`maskBackgroundOpacity`  **- The opacity of the mask layer.** *Default is 0.3*

`shouldKeepSelectedCellColor` **- The boolean value that decides if selected color of cell is visible when the menu is shown.** *Default is false*

`shouldChangeTitleText` **- The boolean value that decides if you want to change the title text when a cell is selected.** *Default is true*

`selectedCellTextLabelColor`  **- The color of the selected cell text label.** *Default is darkGrayColor()*

`arrowTintColor`  **- The tint color of the arrow.** *Default is whiteColor()*

## Requirement
- iOS 9.3+
- Xcode 9.0+, Swift 4.0+

## Changelog
See the [CHANGELOG](https://github.com/gunnarhafdal/GHNavigationDropdownMenu/blob/master/CHANGELOG.md) for details

## License
GHNavigationDropdownMenu is available under the MIT License. See the [LICENSE](https://github.com/gunnarhafdal/GHNavigationDropdownMenu/blob/master/LICENSE) for details.


CHANGELOG.md
## [1.0.4](https://github.com/PhamBaTho/BTNavigationDropdownMenu/releases/tag/2.0.0) (28-09-2017)
- Updated to Swift 4.0
- CellHeight is now a CGFloat instead of NSNumber.

## [1.0.4](https://github.com/PhamBaTho/BTNavigationDropdownMenu/releases/tag/1.0.4) (23-09-2017)
- Updated to Swift 3.1
- Using Double.pi instead of M_PI as per Swift 3.1

## [0.3](https://github.com/PhamBaTho/BTNavigationDropdownMenu/releases/tag/0.3) (17-04-2016)
- Add toggle() functions
- Add keepSelectedCellColor property
- Fixed issues [#41](https://github.com/PhamBaTho/BTNavigationDropdownMenu/issues/41)
- Fixed issues [#38](https://github.com/PhamBaTho/BTNavigationDropdownMenu/issues/38)

## [0.2.1](https://github.com/PhamBaTho/BTNavigationDropdownMenu/releases/tag/0.2.1) (10-01-2016)
- Public show() and hide() functions
- Public isShown property

## [0.2](https://github.com/PhamBaTho/BTNavigationDropdownMenu/releases/tag/0.2) (15-12-2015)
- Passed navigationController when initializing
- Added the alignment of the text inside cell
- Fixed issue [#16](https://github.com/PhamBaTho/BTNavigationDropdownMenu/issues/16): Sometimes navigationController is nil

## [0.1.8](https://github.com/PhamBaTho/BTNavigationDropdownMenu/releases/tag/0.1.8) (15-09-2015)
- Fixed issue: Sometimes navigationController is nil

## [0.1.7](https://github.com/PhamBaTho/BTNavigationDropdownMenu/releases/tag/0.1.7) (04-09-2015)
- Use optional for some properties that relating to navigation controller

## [0.1.6](https://github.com/PhamBaTho/BTNavigationDropdownMenu/releases/tag/0.1.6) (28-08-2015)
- No need to use containerView param when initializing menuView now

## [0.1.5](https://github.com/PhamBaTho/BTNavigationDropdownMenu/releases/tag/0.1.5) (27-08-2015)
- Support device orientation
- Support Split View on iPad

## [0.1.4](https://github.com/PhamBaTho/BTNavigationDropdownMenu/releases/tag/0.1.4) (26-08-2015)
- Fixed issue [#5] (https://github.com/PhamBaTho/BTNavigationDropdownMenu/issues/5). The dropdown overlay appears overtop the tab bar now
- Refactoring and clean up

## [0.1.3](https://github.com/PhamBaTho/BTNavigationDropdownMenu/releases/tag/0.1.3) (25-07-2015)
- Use `usingSpringWithDamping` instead of `contentOffset` for show/hide animation

## [0.1.2](https://github.com/PhamBaTho/BTNavigationDropdownMenu/releases/tag/0.1.2) (21-07-2015)
- Allow change of arrow image after creation of menu


Demo/Demo/AppDelegate.swift
//
//  AppDelegate.swift
//  GHNavigationDropdownMenu
//
//  Created by Pham Ba Tho on 6/8/15.
//  Copyright (c) 2015 PHAM BA THO. All rights reserved.
//

import UIKit

@UIApplicationMain
class AppDelegate: UIResponder, UIApplicationDelegate {

    var window: UIWindow?


    func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplicationLaunchOptionsKey: Any]?) -> Bool {
        return true
    }

    func applicationWillResignActive(_ application: UIApplication) {
        // Sent when the application is about to move from active to inactive state. This can occur for certain types of temporary interruptions (such as an incoming phone call or SMS message) or when the user quits the application and it begins the transition to the background state.
        // Use this method to pause ongoing tasks, disable timers, and throttle down OpenGL ES frame rates. Games should use this method to pause the game.
    }

    func applicationDidEnterBackground(_ application: UIApplication) {
        // Use this method to release shared resources, save user data, invalidate timers, and store enough application state information to restore your application to its current state in case it is terminated later.
        // If your application supports background execution, this method is called instead of applicationWillTerminate: when the user quits.
    }

    func applicationWillEnterForeground(_ application: UIApplication) {
        // Called as part of the transition from the background to the inactive state; here you can undo many of the changes made on entering the background.
    }

    func applicationDidBecomeActive(_ application: UIApplication) {
        // Restart any tasks that were paused (or not yet started) while the application was inactive. If the application was previously in the background, optionally refresh the user interface.
    }

    func applicationWillTerminate(_ application: UIApplication) {
        // Called when the application is about to terminate. Save data if appropriate. See also applicationDidEnterBackground:.
    }


}



Demo/Demo/ViewController.swift
//
//  ViewController.swift
//  GHNavigationDropdownMenu
//
//  Created by Pham Ba Tho on 6/8/15.
//  Copyright (c) 2015 PHAM BA THO. All rights reserved.
//

import UIKit

class ViewController: UIViewController {

    @IBOutlet weak var selectedCellLabel: UILabel!
    var menuView: GHNavigationDropdownMenu!
    
    override func viewDidLoad() {
        super.viewDidLoad()
        let items = [
            GHDropdownItem(title: "Most Popular", subtitle: "from our users"),
            GHDropdownItem(title: "Latest", subtitle: ""),
            GHDropdownItem(title: "Trending", subtitle: "See what's happening"),
            GHDropdownItem(title: "Nearest", subtitle: "to you"),
            GHDropdownItem(title: "Top Picks", subtitle: "by our editors"),
        ]
        self.selectedCellLabel.text = items.first?.title
        self.navigationController?.navigationBar.isTranslucent = false
        self.navigationController?.navigationBar.barTintColor = UIColor(red: 0.0/255.0, green:180/255.0, blue:220/255.0, alpha: 1.0)
        self.navigationController?.navigationBar.titleTextAttributes = [NSAttributedStringKey.foregroundColor: UIColor.white]
        
        
        
        menuView = GHNavigationDropdownMenu(navigationController: self.navigationController, containerView: self.navigationController!.view, selected: items[0], items: items)
        menuView.cellHeight = 50
        menuView.cellBackgroundColor = self.navigationController?.navigationBar.barTintColor
        menuView.cellSelectionColor = UIColor(red: 0.0/255.0, green:160.0/255.0, blue:195.0/255.0, alpha: 1.0)
        menuView.shouldKeepSelectedCellColor = true
        menuView.cellTextLabelColor = UIColor.white
        menuView.cellTextLabelFont = UIFont(name: "Avenir-Heavy", size: 17)
        menuView.cellTextLabelAlignment = .left // .Center // .Right // .Left
        menuView.arrowPadding = 15
        menuView.animationDuration = 0.5
        menuView.maskBackgroundColor = UIColor.black
        menuView.maskBackgroundOpacity = 0.3
        menuView.didSelectItemAtIndexHandler = {(indexPath: Int) -> () in
            print("Did select item at index: \(indexPath)")
            self.selectedCellLabel.text = items[indexPath].title
        }
        
        self.navigationItem.titleView = menuView
    }
}

